File: test/integ/DedupVirtualMethodsTest.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "DedupVirtualMethods.h"
#include "AnnoUtils.h"
#include "RedexTest.h"
#include "Show.h"

class DedupVirtualMethodsTest : public RedexIntegrationTest {};

struct VMethodsStats {
  uint32_t total{0};
  uint32_t annotated{0};
};

VMethodsStats count_methods(const Scope& scope, DexType* annotation) {
  VMethodsStats stats;
  for (auto cls : scope) {
    for (auto method : cls->get_vmethods()) {
      stats.total++;
      stats.annotated +=
          (((get_annotation(method, annotation) == nullptr)) ? 0 : 1);
    }
  }
  return stats;
}

std::vector<DexMethod*> annotated_by_publicized(const Scope& scope) {
  auto publicized_annotation =
      DexType::get_type("Lcom/facebook/redextest/Publicized;");
  std::vector<DexMethod*> result;
  for (auto cls : scope) {
    for (auto method : cls->get_vmethods()) {
      if (get_annotation(method, publicized_annotation)) {
        result.push_back(method);
      }
    }
  }
  return result;
}

void check_public(const std::vector<DexMethod*>& methods,
                  bool should_be_public) {
  for (auto method : methods) {
    if (!method->is_def()) {
      continue;
    }
    EXPECT_EQ(is_public(method), should_be_public)
        << show(method)
        << (should_be_public ? " should be public" : " should not be public");
  }
}

TEST_F(DedupVirtualMethodsTest, dedup) {
  auto scope = build_class_scope(stores);
  auto annotation = DexType::get_type("Lcom/facebook/redextest/Duplication;");

  auto methods_annotated_by_pub = annotated_by_publicized(scope);
  check_public(methods_annotated_by_pub, false);

  auto before_stats = count_methods(scope, annotation);
  auto deduplicated_vmethods = dedup_vmethods::dedup(stores);
  auto after_stats = count_methods(scope, annotation);

  EXPECT_EQ(after_stats.annotated, 0);
  EXPECT_EQ(deduplicated_vmethods, before_stats.annotated);
  EXPECT_EQ(before_stats.total - before_stats.annotated, after_stats.total);
  check_public(methods_annotated_by_pub, true);
}

File: test/integ/DedupVirtualMethodsTest.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.redextest;

import java.lang.annotation.Target;
import java.lang.annotation.ElementType;

/**
 * All the methods that are annotated by Duplication will be removed.
 */
@Target(ElementType.METHOD)
@interface Duplication {
}

/**
 * All the methods that are annotated by Duplication will be publicized.
 */
@Target(ElementType.METHOD)
@interface Publicized {
}

class P {
  public int method() {
    return 0;
  }
  @Publicized
  protected int non_public_method() {
    return 1;
  }
}

class C1 extends P {}
class C11 extends C1 {
  @Duplication
  public int method() {
    return 0;
  }
  @Duplication
  public int non_public_method() {
    return 1;
  }
}
class C111 extends C11 {
  public int method() {
    return 1;
  }
}
class C1111 extends C111 {
  public int method() {
    return 0;
  }
}

class C2 extends P {
  @Duplication
  public int method() {
    return 0;
  }
  // This is not deduplicated but its access flag should be public after the transformation.
  @Publicized
  protected int non_public_method() {
    return 2;
  }
}
class C21 extends C2 {
  @Duplication
  public int method() {
    return 0;
  }
}

class PD {
  public int method() {
    return 0;
  }
}

class C2D extends PD {
  @Duplication
  public int method() {
    return 0;
  }

  void callV() {
    C21D c = new C21D();
    c.method();
  }
}
class C21D extends C2D {
  @Duplication
  public int method() {
    return 0;
  }
}
